### ISSUE TITLE
Transfer learning Issue

### ISSUE THREAD

**Issue by Saiflalwani** [ultralytics/ultralytics#2330]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

i have trained yolo v8 model for englishcards having classes ('Smart Identifier',' Smart Back', 'Logo','Picture',' Smart Front','Damaged')

this is the yaml file 
train: ../train/images
val: ../valid/images
test: ../test/images

nc: 6
names: [' Smart Identifier',  'Smart Back',   ' Logo',   'Picture',  'Smart Front',  'Damaged']
the result is save in /content/detect/train/weights/best.pt

now i have seprately annotated urducard having classes  ('urdu Identifier',' urduBack', 'Logo','Picture',' urduFront')  the sequence of annotation ('urdu Identifier',' urduBack', 'Logo','Picture',' urduFront')    starting from 0 index 
so urdu have 5 classes  i whant to consolidate these twoenglishcard and urducard    i both models Logo and picture are same 

so i have created a new yaml file 
train: ../train/images
val: ../valid/images
test: ../test/images

nc: 11
names: [ ' Urdu Identifier', ' Urdu Back',   'Logo',   'Picture'  ,' Urdu Front'' Identifier' ,'Smart Back',  ' Logo',  'Picture' ,'Smart Front',   'Damaged']

i have combined all 11 classes   

i have used this as my training code 

!yolo task=detect mode=train model=/content/detect/train/weights/best.pt   data=new_data.yaml epochs=100 batch=16 imgsz=650  plots=True hsv_h=0.015 hsv_s=0.5 hsv_v=0.3 degrees=0.0 translate=0.1 scale=0.2 shear=1.0 flipud=0.5 fliplr=0.5  

issue is it is not able to detect  the classes in pretrained model (classes in englishcards)
how to resolve this issue

 

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@Saiflalwani hello! Based on the description of your issue, it seems like you have trained a YOLOv8 model for English cards with six classes, and have separately annotated Urdu cards with five classes but with some overlapping classes. You want to consolidate these two datasets into one model with 11 classes. However, when you use the new combined dataset to train the model, it is not able to detect the classes from the original English cards dataset. 

To address this issue, one potential solution is to use transfer learning. You can use the weights from your pre-trained YOLOv8 model on the English cards dataset as a starting point, and then continue training on the combined dataset of English and Urdu cards. This will help the model learn to detect the additional classes while still retaining the ability to detect the original English classes. To do this, you can use the `transfer` flag in your training command to load in the weights from the best.pt file of your English cards model. Additionally, you may need to adjust some of the training hyperparameters such as the learning rate or the number of epochs to better balance the new classes.

I hope this helps. Let me know if you have any further questions or issues!

**Comment by Saiflalwani** [comment]
can we use transfer flag in yolo v8 ?

is this correct syntex
!yolo task=detect mode=train model=/content/detect/train/weights/best.pt data=data.yaml epochs=100 batch=16 imgsz=650 pretrained=True  --transfer

**Comment by glenn-jocher** [comment]
@Saiflalwani Yes, in YOLOv8 you can use the `--transfer` flag to enable transfer learning. In your command, the `--pretrained` flag should already load in the pre-trained weights, so you do not need to specify `model=/content/detect/train/weights/best.pt` separately. 

The correct syntax would be:
`!yolo task=detect mode=train data=new_data.yaml epochs=100 batch=16 imgsz=650 pretrained=True --transfer`

If you are still experiencing an error, you may need to provide more information on the specific error message you are receiving.

**Comment by Saiflalwani** [comment]
@glenn-jocher 
i am using this traning code genrating error 
!yolo task=detect mode=train data=new_data.yaml epochs=100 batch=16 imgsz=650 pretrained=True --transfer


WARNING ⚠️ '--transfer' does not require leading dashes '--', updating to 'transfer'.
Traceback (most recent call last):
  File "/usr/local/bin/yolo", line 8, in <module>
    sys.exit(entrypoint())
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/yolo/cfg/__init__.py", line 332, in entrypoint
    check_cfg_mismatch(full_args_dict, {a: ''})
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/yolo/cfg/__init__.py", line 178, in check_cfg_mismatch
    raise SyntaxError(string + CLI_HELP_MSG) from e
SyntaxError: 'transfer' is not a valid YOLO argument. Similar arguments are i.e. ['translate=0.1', 'tracker=botsort.yaml', 'train'].

    Arguments received: ['yolo', 'task=detect', 'mode=train', 'data=data.yaml', 'epochs=100', 'batch=16', 'imgsz=650', 'pretrained=True', '--transfer']. Ultralytics 'yolo' commands use the following syntax:

        yolo TASK MODE ARGS

        Where   TASK (optional) is one of ('detect', 'segment', 'classify', 'pose')
                MODE (required) is one of ('train', 'val', 'predict', 'export', 'track', 'benchmark')
                ARGS (optional) are any number of custom 'arg=value' pairs like 'imgsz=320' that override defaults.
                    See all ARGS at https://docs.ultralytics.com/usage/cfg or with 'yolo cfg'

![Capture](https://user-images.githubusercontent.com/114072388/235372975-6e15a2ae-3bec-4056-99eb-5846be3599bd.JPG)
this is the weight i want to use i imported this complete weight which was trained earlier with classes as    'Smart Identifier',' Smart Back', 'Logo','Picture',' Smart Front','Damaged'  

now i have separately annotates urdu sets containg  urdu Identifier',' urduBack', 'Logo','Picture',' urduFront'                                                  
now i want apply transfer learning so i can make a consolidate model 
my yaml file 
train: ../train/images
val: ../valid/images
test: ../test/images

nc: 11
names: [ ' Urdu Identifier', ' Urdu Back', 'Logo', 'Picture' ,' Urdu Front'' Identifier' ,'Smart Back', ' Logo', 'Picture' ,'Smart Front', 'Damaged']

i have uses this code for training
!yolo task=detect mode=train model=/content/detect/train/weights/best.pt data=data.yaml epochs=100 batch=16 imgsz=650 pretrained=True

model=/content/detect/train/weights/best.pt
this is the model that is trained on englishcards 
in data.yaml training images contain only urdu datasets that is annotated on urdu classes 

what things to be updated in training and yaml file what else should i do to complete this task

**Comment by glenn-jocher** [comment]
@Saiflalwani It seems like you are getting a syntax error when trying to use the `--transfer` flag in your training command. The syntax error is because `--transfer` is not a valid YOLOv8 argument. To use the transfer learning in YOLOv8, you should simply use the `transfer=True` argument instead. So, the correct command syntax would be:
`!yolo task=detect mode=train data=new_data.yaml epochs=100 batch=16 imgsz=650 pretrained=True transfer=True`

Additionally, since you are performing transfer learning in YOLOv8, you should not use the original `data.yaml` file containing the classes from the English cards dataset. Instead, you should use the new `new_data.yaml` file you created containing the consolidated classes from both datasets.

Lastly, as a general suggestion, you may want to adjust some hyperparameters such as the learning rate or scale to help the model better learn the new classes while minimizing overfitting. 

I hope this helps. Let me know if you have any further questions or issues!

**Comment by Saiflalwani** [comment]
Error is coming
Traceback (most recent call last):
  File "/usr/local/bin/yolo", line 8, in <module>
    sys.exit(entrypoint())
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/yolo/cfg/__init__.py", line 335, in entrypoint
    check_cfg_mismatch(full_args_dict, overrides)
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/yolo/cfg/__init__.py", line 178, in check_cfg_mismatch
    raise SyntaxError(string + CLI_HELP_MSG) from e
SyntaxError: 'transfer' is not a valid YOLO argument. Similar arguments are i.e. ['translate=0.1', 'tracker=botsort.yaml', 'train'].

    Arguments received: ['yolo', 'task=detect', 'mode=train', 'data=data.yaml', 'epochs=100', 'batch=16', 'imgsz=650', 'pretrained=True', 'transfer=True']. Ultralytics 'yolo' commands use the following syntax:

        yolo TASK MODE ARGS

        Where   TASK (optional) is one of ('detect', 'segment', 'classify', 'pose')
                MODE (required) is one of ('train', 'val', 'predict', 'export', 'track', 'benchmark')
                ARGS (optional) are any number of custom 'arg=value' pairs like 'imgsz=320' that override defaults.
                    See all ARGS at https://docs.ultralytics.com/usage/cfg or with 'yolo cfg'



same error is coming for transfer=True 
do i need to include  freeze

**Comment by glenn-jocher** [comment]
@Saiflalwani hello! 

The error message you are receiving is because `transfer` is not a valid YOLOv8 argument. To enable transfer learning, you should use the `transfer=True` argument instead. So, the correct command syntax would be:
`yolo task=detect mode=train data=data.yaml epochs=100 batch=16 imgsz=650 pretrained=True transfer=True`

You do not need to include `freeze` for transfer learning. The `freeze` argument is used to freeze certain layers of the model during training, which can help improve its performance for certain use cases.

I hope this helps! Let me know if you have any further questions or issues.

**Comment by PasinduShavinda** [comment]
@glenn-jocher Can you please provide the tutorial for refer and get understanding about Yolov8 Transfer Learning..

**Comment by glenn-jocher** [comment]
@PasinduShavinda hello! Thank you for your interest in YOLOv8 transfer learning. While there may not be a specific tutorial dedicated to transfer learning in YOLOv8, you can refer to the official YOLOv8 documentation for detailed information on how to perform transfer learning.

In transfer learning with YOLOv8, you would typically start with a pre-trained model on a large dataset (such as COCO) and fine-tune it on your specific dataset. To do this, you would need to update the data configuration file (`data.yaml`) to include your own classes and paths to your training images, validation images, and test images.

Then, you can use the `task=detect` and `mode=train` arguments along with other necessary parameters such as `epochs`, `batch size`, and `img size` to initiate the transfer learning process. Including the `pretrained=True` argument is important to load the pre-trained weights into the model.

It's worth noting that you should ensure your dataset follows the same object detection format as the original YOLOv8 implementation, with bounding box annotations and corresponding labels.

If you have any specific questions or need further guidance during the transfer learning process, please feel free to ask.

**Comment by PasinduShavinda** [comment]
@glenn-jocher Thank You Very Much for your clear explenation !,  I have a small question regarding YoloV8 Transfer learning, I'm using pre-trained model on a large dataset (COCO), I notice that the COCO dataset image size is 640×480, In my dataset image size was 416x416 will it be an issue when doing transfer learning by fine-tune it on my specific dataset?

**Comment by glenn-jocher** [comment]
@PasinduShavinda Thank you for your question! The difference in image sizes between the COCO dataset (640x480) and your specific dataset (416x416) should not pose a major issue when performing transfer learning with YOLOv8.

During transfer learning, the pre-trained model learns high-level features from the large COCO dataset. These features capture general object representations that can be useful for detecting objects in different domains, including your specific dataset.

When you fine-tune the pre-trained model on your dataset with smaller images (416x416), the model will adapt and learn the specific features and object representations relevant to your dataset. The downsized images may even help to capture more detailed information about the objects.

However, it's important to note that using a smaller image size could result in a trade-off between detection accuracy and computation speed. Smaller images might yield lower accuracy compared to larger images, but they can also lead to faster inference times. You may need to experiment with different image sizes to find the optimal balance for your specific use case.

Overall, the slight difference in image sizes between COCO and your dataset should not be a significant issue for transfer learning, and you can fine-tune the pre-trained model on your dataset effectively.

Please let me know if you have any further questions or concerns!

**Comment by PasinduShavinda** [comment]
@glenn-jocher I have a question regarding YoloV8 Transfer learning. If I do the freeze backbone of the model and train the last fully connected layers with my custom dataset will be improve the training accuracy ?

**Comment by glenn-jocher** [comment]
@PasinduShavinda hi there,

When performing transfer learning with YOLOv8, freezing the backbone of the model and only training the last fully connected layers with your custom dataset can potentially improve training accuracy. The backbone usually consists of pre-trained layers that have been trained on a large dataset (such as COCO) and have learned general object representations. By freezing these layers, you retain these learned representations and avoid overwriting them during training.

By focusing the training on the last fully connected layers, you allow the model to specialize and learn specific features relevant to your custom dataset. This fine-tuning process can help the model adapt to your data and improve the accuracy of object detection.

However, it's important to note that the effectiveness of freezing the backbone and training only the last layers can vary depending on the complexity and similarity of your custom dataset to the pre-training dataset. In some cases, it may be beneficial to fine-tune some of the pre-trained backbone layers to achieve better performance.

It's always a good idea to experiment with different approaches and see what works best for your specific dataset and use case. You can try freezing the backbone and training the last layers, and also consider fine-tuning some of the backbone layers if necessary.

I hope this explanation helps! If you have any further questions, feel free to ask.

**Comment by PasinduShavinda** [comment]
@glenn-jocher Thank You for clear explanation.. I have a question regarding YoloV8 backbone architecture, How can I find the backbone implementation code file in ultralytics github repository ?

**Comment by glenn-jocher** [comment]
@PasinduShavinda hello! Thank you for your question. To find the implementation code file for the YOLOv8 backbone architecture in the ultralytics GitHub repository, you can navigate to the `models` directory. Inside this directory, you will find the `yolo.py` file which contains the implementation of the YOLOv8 model architecture, including the backbone.

Within the `yolo.py` file, you can search for the relevant sections or functions that define the backbone of the YOLOv8 model. The backbone typically consists of convolutional layers that extract features from the input image and form the foundation of the model's object detection capabilities.

By examining the code within `yolo.py`, you can gain a deeper understanding of the backbone architecture used in YOLOv8 and how it is implemented in the ultralytics GitHub repository.

I hope this helps! Let me know if you have any further questions or need additional clarification regarding the YOLOv8 backbone architecture.

**Comment by RajUpadhyay** [comment]
@glenn-jocher 
Hi,
I tried your command but it still gives the error transfer is not a valid YOLO command
  ```yolo task=detect mode=train data=data.yaml epochs=100 batch=16 imgsz=650 pretrained=True transfer=True```

**Comment by sachinlodhi** [comment]
> @RajUpadhyay hi there,
> 
> I see you're having an issue with running the `yolo` command with the `transfer` argument.
> 
> The `transfer` argument is not valid in YOLOv8. Instead, transfer learning in YOLOv8 is performed by loading a pre-trained model with the argument `pretrained=True`. This will initialize your model with pre-trained weights which you can then fine-tune on your specific dataset.
> 
> To conduct transfer learning in YOLOv8, you should adjust the other arguments as appropriate (task, mode, data, epochs, batch, imgsz) without the `transfer` argument.
> 
> I hope this information helps! If you have any other questions, feel free to ask.

Hey does it mean if I am training the model initially using pretrained=True then all 80 base classe objects will be detected at the end of the training? Because I have used the following code:
`!yolo task=detect \
mode=train \
data=/content/pet_plastic-1/data.yaml \
epochs=10 \
imgsz=640 pretrained=True`
 
 and my data.yaml looks like:
 `names:
  0: person
  1: bicycle
  2: car
  3: motorcycle
  4: airplane
  5: bus
  6: train
  7: truck
  8: boat
  9: traffic light
  10: fire hydrant
  11: stop sign
  12: parking meter
  13: bench
  14: bird
  15: cat
  16: dog
  17: horse
  18: sheep
  19: cow
  20: elephant
  21: bear
  22: zebra
  23: giraffe
  24: backpack
  25: umbrella
  26: handbag
  27: tie
  28: suitcase
  29: frisbee
  30: skis
  31: snowboard
  32: sports ball
  33: kite
  34: baseball bat
  35: baseball glove
  36: skateboard
  37: surfboard
  38: tennis racket
  39: bottle
  40: wine glass
  41: cup
  42: fork
  43: knife
  44: spoon
  45: bowl
  46: banana
  47: apple
  48: sandwich
  49: orange
  50: broccoli
  51: carrot
  52: hot dog
  53: pizza
  54: donut
  55: cake
  56: chair
  57: couch
  58: potted plant
  59: bed
  60: dining table
  61: toilet
  62: tv
  63: laptop
  64: mouse
  65: remote
  66: keyboard
  67: cell phone
  68: microwave
  69: oven
  70: toaster
  71: sink
  72: refrigerator
  73: book
  74: clock
  75: vase
  76: scissors
  77: teddy bear
  78: hair drier
  79: toothbrush
  80: NEW_OBJECT
roboflow:
  license: CC BY 4.0
  project: object-detection-lol
  url: https://universe.roboflow.com/repo/repo/dataset/1
  version: 1
  workspace: aladeenmonk39
test: ../test/images
train: ../train/images
val: ../valid/images
`
 here class index 80 is the new object I am training the model again.
 After traning the model using the command mentioned above and when I test it then it doesn't detect all other 80 objects except the one I just trained it for. Where am I making mistake?

**Comment by glenn-jocher** [comment]
Hi @sachinlodhi, 

The behavior you're observing is actually expected. When you fine-tune a pre-trained model with your custom dataset (which in your case includes a new class at index 80), the model will primarily learn to detect the classes in your custom dataset during the fine-tuning process.

In effect, you're adapting the model to specialize in detecting the objects in your new dataset. Because of this, you may see degraded performance in detecting the original set of classes that the pre-trained model was initially trained on. This is because the fine-tuning process can cause the model to somewhat "forget" the old classes as it learns to detect the new ones.

In your case, if you want to detect all 80 base classes of objects plus your new object(s), you would ideally want to train with a dataset that includes examples of all these object classes. However, if you only have data for your new object class, a potential approach is to mix in some portion of the original dataset (i.e., COCO) that covers the 80 base classes during training, to maintain the model's ability to detect those classes as well.

Hope this helps clarify your question. Let me know if you need any more assistance!

**Comment by sachinlodhi** [comment]
Thanks for your repsonse. I dont want to train the model with 80 classes. Now my objective is to train the model for just two classes. Now the problem is that I am trying to traing the model twice for 2 different classes. 
My first data1.yaml file looks like this:

`names:
- object_A
nc: 1
roboflow:
  license: CC BY 4.0
  project: object_1_detection
  url: https://universe.roboflow.com/dataDIR/dataset/1
  version: 1
  workspace: abracadabra
test: ../test/images
train: ../train/images
val: ../valid/images
`

Then I train the model using following command(for first object detection):
`!yolo task=detect \
mode=train \
model=yolov8n.pt   \
data=/content/data1/data1.yaml \
epochs=100 \
imgsz=640 
`
after training completes it is able to detect the object for the class object_A so not I take another dataset for object_B and its data2.yaml looks like:
`
nc: 2
names: ['object_B','object_A']
roboflow:
  license: CC BY 4.0
  project: dataB
  url: ttps://universe.roboflow.com/dataset_B/dataset/2
  version: 2
  workspace: workspace2
test: dataset/test/images
train: dataset/train/images
val: dataset/koreatech-all-2/valid/images
`
and now when I run this using the follwoing command:

`
!yolo task=detect \
mode=train \
model=/content/runs/detect/train/weights/best.pt \
data=/content/dataset2/data2.yaml \
epochs=20 \
imgsz=640 pretrained=True
`

now the training is successful for second dataset too and detecting the objects of the class object_B but it does not detect the objects from class object_A.

If summerise in short:
I trained the model for  dataset1, modified the data2.yaml file as `nc: 2
names: ['object_B', 'object_A']`
then retrained the model dataset2 with pretrained=True flag and and after the training it can identify the objects from the dataset of the last training only and not before that. 
what can be the issues?

**Comment by glenn-jocher** [comment]
Hello @sachinlodhi,

Thanks for sharing more details, it's very helpful to fully understand your situation. In object detection models like YOLOv8, the learnt parameters are quite specific to the classes involved during the training process.

When you are training your model for the first time with just 'object_A', the model learns to detect instances of 'object_A'. Then, when you train the model again with another dataset and a different .yaml file that includes 'object_B' and 'object_A', you're essentially overriding the previous knowledge of the model. This is why it does not detect 'object_A' anymore, despite it being included in the list of classes for the second training.

When you use the `pretrained=True` parameter, you are initializing your model with pre-trained weights. However, during the training process, these weights get updated. Even though 'object_A' is listed in the second .yaml file, unless there are actual instances of 'object_A' in your second dataset, the model will not effectively learn to detect 'object_A'.

In short, if you want your model to detect both 'object_A' and 'object_B', you should ideally train your model on a dataset that includes instances of both these classes. This will help ensure the model learns the necessary features for detecting both 'object_A' and 'object_B'. If your datasets for 'object_A' and 'object_B' are separate, consider creating a combined dataset that includes both classes and use this combined dataset for training.

I hope this explanation helps! Let me know if you have more questions.
